api/wft-complete-metadata.md

```diff
@@ -0,0 +1,126 @@
+# Workflow Task Completion Metadata
+
+We have a need to attach various bits of metadata to workflow task completions. There are a number
+of usecases in which the SDK would like to save some information for itself, or tell server about
+something, but this information does not influence replay or the reconstitution of workfow state in
+any way. Thus, it shouldn't go in a marker or some new kind of history event, because this would
+violate the (useful) idea that history events are primarily concerned with recreating workflow
+state.
+
+## Use cases
+
+### Reporting currently-being-awaited commands & facts about them
+
+Right now, besides a query, there's no simple way for a user to understand what their workflow is
+actively waiting on. Solving this would make it much easier for users to have insight into why
+their workflow is or isn't making progress, and could be nicely exposed in the UI.
+
+In particular, understanding what's going on while your workflow is currently running local
+activities and performing WFT heartbeats is very opaque. This would make that much more clear.
+
+Imagine being able to get information like this (and do useful things with it in the UI):
+
+```
+waiting_on: [
+    { command_type: "local_activity", activity_name: "la_func", id: 1, retries: 0, location: "my_file.py:23" },
+    { command_type: "local_activity", activity_name: "la_func", id: 2, retries: 5, location: "my_file.py:23" },
+    { command_type: "timer", id: 3, duration: "5s", location: "my_file.py:42" },
+]
```

Comment:
Will those payload be encoded by user's custom data converter? I would assume some of them would not if server needs to understand them, but others would because it could be sensitive info.

Reply:
This specific example of outstanding commands should use the standard data converter. They may need to be understood for cloud billing purposes (the local activity stuff, depending on our choices there). They also need to be UI-renderable to be useful for debugging info.

But, regardless, again, that's not really required to move forward with this proposal. The specifics of any piece of metadata is up to if and when we add it. This proposal is just the mechanism of transport.